File: Sorting/75_sort_colors.cpp
#include <vector>
#include <iostream>
#include <sstream>


// O(N)
class Solution {
public:
    void sortColors(std::vector<int>& nums) {
        std::vector<int> right_most = {-1, -1, -1};
        
        for (int i = 0; i < nums.size(); i++){
            if (nums[i] == 0) {
                std::swap(nums[right_most[0] + 1], nums[i]);
                // repeating code for readability, to separate 0/1/2 cases
                if (nums[i] == 1)
                    std::swap(nums[right_most[1] + 1], nums[i]);
                right_most[0]++; right_most[1]++; right_most[2]++;
            }
            else if (nums[i] == 1){
                std::swap(nums[right_most[1] + 1], nums[i]);
                right_most[1]++; right_most[2]++;
            }
            else{
                right_most[2]++;
            }
        }
        
    }
};


int main(){
    std::cout << "Space separated colors: ";
    std::string line;
    std::getline(std::cin, line);
    
    // split by space
    std::vector<int> nums;
    int c;
    std::istringstream iss(line);
    while (iss >> c) {    
        nums.push_back(c);
    }
    
    Solution().sortColors(nums);
    std::cout << "Sorted colors: ";
    for (auto v: nums) std::cout << v << " ";
    std::cout << std::endl;

    return 0;
}

File: Sorting/75_sort_colors.py
# O(N)
class Solution(object):
    def sortColors(self, nums) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        right_most = [-1, -1, -1]
        
        for i in range(len(nums)):
            if nums[i] == 0:
                nums[i], nums[right_most[0] + 1] = nums[right_most[0] + 1], nums[i]
                if nums[i] == 1:
                    nums[i], nums[right_most[1] + 1] = nums[right_most[1] + 1], nums[i]
                right_most[0] += 1
                right_most[1] += 1
                right_most[2] += 1
            elif nums[i] == 1:
                nums[i], nums[right_most[1] + 1] = nums[right_most[1] + 1], nums[i]
                right_most[1] += 1
                right_most[2] += 1
            else:
                right_most[2] += 1


if __name__ == '__main__':
    nums = input('Space separated colors: ')
    nums = [int(v) for v in nums.split(' ')]
    Solution().sortColors(nums)
    print('Sorted colors: ', nums)